zipkin-server/src/main/java/zipkin2/server/internal/elasticsearch/DynamicCredentialsFileLoader.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2015-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.server.internal.elasticsearch;
+
+import com.google.common.util.concurrent.RateLimiter;
```

Comment:
yeah no need for a dependency to do this. If I understand correctly, this is a single-thread that's updating things (same thread). you can implement an interval without any dep or fancy using System.nanoTime and subtraction. 

```java
MyRunnable {
  final SameThreadRateLimiter logLimiter = new SameThreadRateLimiter(5);

  public void run(){
    try {
         // do fancy thing
    } catch (Throwable e) {
       Call.propagateIfFatal(e);
       if (logLimiter.pass()) {
          // LOG it!
       }
    }
  }

SameThreadRateLimiter {
  static final long NANOS_PER_SECOND = TimeUnit.SECONDS.toNanos(1);
  final int intervalSeconds;
  long nextPass = nextPass(System.nanoTime()); // pass initially

  SameThreadRateLimiter(int intervalSeconds) {
    this.intervalSeconds = intervalSeconds;
  }

  boolean pass() {
    return shouldPass(System.nanoTime());
  }

  static long nextPass(long now) {
    return now + intervalSeconds * NANOS_PER_SECOND;
  }

  static boolean shouldPass(long now) {
    long nanosUntilPass = -(now - nextPass); // because nanoTime can be negative
    if (nanosUntilPass <= 0) {
      nextPass = nextPass(now);
      return true;
    }
    return false;
    }
}
```

logic simplified due to same thread from this 
https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/sampler/RateLimitingSampler.java

if you use this, please add unit test on static method, which you can mostly copy paste  from 
https://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/sampler/RateLimitingSamplerTest.java

Reply:
Is this to reduce logging if a user sets a really small refresh interval? It seems like overkill especially since most will be in the seconds, how about just logging without the limit?

Reply:
I switch to log error without any rate limit.